### ISSUE TITLE
Keras model.compile(..., metrics=["accuracy"]) no longer introspects loss function in TF 2.2

### ISSUE THREAD

**Issue by huonw** [tensorflow/tensorflow#41361]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS, Colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): pip/binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b, 2.2.0
- Python version: 3.7.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**

The documentation of [`tf.keras.Model.compile`](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) includes the following for the `metrics` parameter:

> When you pass the strings 'accuracy' or 'acc', we convert this to one of tf.keras.metrics.BinaryAccuracy, tf.keras.metrics.CategoricalAccuracy, tf.keras.metrics.SparseCategoricalAccuracy based on the loss function used and the model output shape. We do a similar conversion for the strings 'crossentropy' and 'ce' as well. 

The code in question only looks at the model shape, and ignores the loss function:

https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/engine/compile_utils.py#L447-L453

This means that using `acc`/`accuracy` for a model doing binary classification with last dimension > 1 (e.g. binary predictions about multiple independent values for each batch element) will incorrectly use categorical accuracy, not binary accuracy. (That is, the `compile` invocation is equivalent to passing `metrics=["categorical_accuracy"]`.)

**Describe the expected behavior**

The behaviour stated by the documentation is to look at the loss function in addition to the output shape. That is, if the loss function is binary cross-entropy, `metrics=["accuracy"]` should be equivalent to `metrics=["binary_accuracy"]`. 

This is behaviour of TF < 2.2, such as TF 2.1.1:

https://github.com/tensorflow/tensorflow/blob/3ffdb91f122f556a74a6e1efd2469bfe1063cb5c/tensorflow/python/keras/engine/training_utils.py#L1114-L1121

**Standalone code to reproduce the issue**

```python
#%pip install tensorflow==2.2.0
#%pip install tensorflow==2.1.1
import tensorflow as tf

inp = tf.keras.Input(3)
model = tf.keras.Model(inp, inp)

model.compile(loss="bce", metrics=["acc"])
model.evaluate(tf.constant([[0.1, 0.6, 0.9]]), tf.constant([[0, 1, 1]]))

if tf.version.VERSION == "2.2.0":
    print(model.compiled_metrics.metrics[0]._fn.__name__)
else:
    print(model._per_output_metrics[0]["acc"]._fn.__name__)
```

Output

- TF 2.2.0: `categorical_accuracy`
- TF 2.1.1: `binary_accuracy`

Notebook: https://gist.github.com/huonw/4a95b73e3d8a1c48a8b5fc5297d30772
Colab: https://colab.research.google.com/gist/huonw/4a95b73e3d8a1c48a8b5fc5297d30772

**Other info / logs** N/A

**Comment by ravikyram** [comment]
I have tried in colab with TF 2.2 and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/cb15be043d402e051ffd25615f7a9291/untitled144.ipynb).Thanks!

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF 2.6.0-dev20210530,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/90978bade8bf4a1016bfe3e82d0ebe70/untitled134.ipynb)..Thanks !

**Comment by bhack** [comment]
> Was able to replicate the issue in TF 2.6.0-dev20210530,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/90978bade8bf4a1016bfe3e82d0ebe70/untitled134.ipynb)..Thanks !

There was a linked PR that was closed without any feedback.

**Comment by bhack** [comment]
Can you check https://github.com/keras-team/keras/pull/16114

**Comment by sachinprasadhs** [comment]
Could you please refer the fix provided in the PR https://github.com/keras-team/keras/pull/16893 which is successfully merged and close this issue. Thanks!
